src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -336,11 +359,75 @@ public void CopyChangedValues(TStructuralType original)
                     bool isDeltaType = TypedDelta.IsDeltaOfT(deltaNestedResource.GetType());
                     Contract.Assert(isDeltaType, nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
 
+                    Type nestedType = deltaNestedResource.StructuredType;
+                    Type originalType = originalNestedResource.GetType();
+
+                    if (deltaNestedResource.IsComplexType && !nestedType.IsAssignableFrom(originalType))
+                    {
+                        originalNestedResource = ReAssignComplexDerivedType(original, nestedResourceName, originalNestedResource, nestedType, originalType);
+                    }
+
                     deltaNestedResource.CopyChangedValues(originalNestedResource);
                 }
             }
         }
 
+        private dynamic ReAssignComplexDerivedType(TStructuralType original, string nestedResourceName, dynamic originalNestedResource, Type nestedType, Type originalType)
+        {
+            //As per OASIS discussion, changing a complex type from 1 derived type to another is allowed if both derived type have a common ancestor and the property
+            //is declared in terms of a common ancestor. The logic below checks for a common ancestor. Create a new object of the derived type in delta request.
+            //And copy the common properties.
+
+            Type complexBaseType = nestedType.BaseType;
+            HashSet<Type> lstComplexBaseTypes = new HashSet<Type>();
+
+            while (complexBaseType != null)
+            {
+                lstComplexBaseTypes.Add(complexBaseType);
+                complexBaseType = complexBaseType.BaseType;
+            }
+
+            //Here original type is the type for original (T) resource.
+            //We will keep  going to base types and finally will get the Common Basetype for the derived complex types in to the originalType variable.
+                        
+            bool foundCommonbase = false;
+
+            //The new Original type, means the new complex type (T) which will replace the current complex type.
+            dynamic newOriginalNestedResource = null;
+
+            while (originalType != null)
```

Comment:
what if we don't find a common ancestor? or do we assert that there should be a common ancestor, since they should both derive from the type of the property?

Reply:
If there is not in cae of deserialization it throws error, in case of user creation this will be skipped and it will throw err at old place where it used to